From 8a25a10fd364dfc59cb674d5433f9b7fd2a051d1 Mon Sep 17 00:00:00 2001
From: Reilly Brogan <reilly@reillybrogan.com>
Date: Fri, 2 Feb 2024 21:16:59 -0600
Subject: [PATCH 1/3] Use KSM for the BSS segment of shared libraries

Shared libraries often have a BSS segment (zero'd variables) that is
often the same runtime content between binaries that use the  library.

This patch turns on KSM specifically for this segment
---
 elf/dl-map-segments.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/elf/dl-map-segments.h b/elf/dl-map-segments.h
index 30977cf800..eb0587ea27 100644
--- a/elf/dl-map-segments.h
+++ b/elf/dl-map-segments.h
@@ -188,6 +188,7 @@ _dl_map_segments (struct link_map *l, int fd,
                               -1, 0);
               if (__glibc_unlikely (mapat == MAP_FAILED))
                 return DL_MAP_SEGMENTS_ERROR_MAP_ZERO_FILL;
+              __madvise((caddr_t) zeropage, zeroend - zeropage, MADV_MERGEABLE);
             }
         }
 
-- 
2.43.0

